unison-cli/src/Unison/CommandLine/Completion.hs

```diff
@@ -0,0 +1,321 @@
+{-
+   This module defines tab-completion strategies for entering info via the CLI
+-}
+module Unison.CommandLine.Completion
+  ( -- * Completers
+    exactComplete,
+    prefixCompleteTermOrType,
+    prefixCompleteTerm,
+    prefixCompleteType,
+    prefixCompletePatch,
+    noCompletions,
+    prefixCompleteNamespace,
+    -- Exported for testing
+    prefixCompletionFilter,
+    -- Unused for now, but may be useful later
+    prettyCompletion,
+    fixupCompletion,
+  )
+where
+
+import Control.Lens (ifoldMap)
+import qualified Control.Lens as Lens
+import Data.List (isPrefixOf)
+import qualified Data.List as List
+import Data.List.Extra (nubOrd)
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map as Map
+import Data.Set.NonEmpty (NESet)
+import qualified Data.Set.NonEmpty as NESet
+import qualified Data.Text as Text
+import qualified System.Console.Haskeline as Line
+import System.Console.Haskeline.Completion (Completion)
+import qualified U.Codebase.Branch as V2Branch
+import qualified U.Codebase.Causal as V2Causal
+import qualified U.Codebase.Reference as Reference
+import qualified U.Codebase.Referent as Referent
+import qualified U.Util.Monoid as Monoid
+import Unison.Codebase (Codebase)
+import qualified Unison.Codebase as Codebase
+import qualified Unison.Codebase.Branch as Branch
+import qualified Unison.Codebase.Path as Path
+import qualified Unison.Codebase.SqliteCodebase.Conversions as Cv
+import qualified Unison.Hash as H
+import qualified Unison.HashQualified' as HQ'
+import qualified Unison.NameSegment as NameSegment
+import Unison.Prelude
+import qualified Unison.ShortHash as SH
+import qualified Unison.Util.Pretty as P
+import Prelude hiding (readFile, writeFile)
+
+-- | Things which we may want to complete for.
+data CompletionType
+  = NamespaceCompletion
+  | TermCompletion
+  | TypeCompletion
+  | PatchCompletion
+  deriving (Show, Eq, Ord)
+
+-- | The empty completor.
+noCompletions ::
+  Monad m =>
+  String ->
+  Codebase m v a ->
+  Branch.Branch m ->
+  Path.Absolute ->
+  m [System.Console.Haskeline.Completion.Completion]
+noCompletions _ _ _ _ = pure []
+
+-- |
+--
+-- Finds names of the selected completion types in the current namespace.
+-- The caller is responsible for filtering those names according to whatever strategy they
+-- want, e.g. prefix matching, fuzzy finding, etc.
+--
+-- Given a codebase with these terms:
+--
+-- @@
+-- .base.List.map.doc
+-- .base.List
+-- .bar.foo
+-- @@
+--
+-- We will return:
+--
+-- @@
+-- -- Note how we return `base` even if it's not a prefix match, it's the caller's job to
```

Comment:
Fix this comment to indicate that this is what's passed to the user filter, not actually returned.